### PR TITLE
Bugfix/corbett/some leaks

### PR DESCRIPTION
**Pull request by corbett5**

### CODE REVIEW

File: src/umpire/strategy/AllocationTracker.cpp [hunk 1]
```diff
@@ -20,13 +20,23 @@ namespace strategy {
 AllocationTracker::AllocationTracker(
   const std::string& name,
   int id,
-  Allocator allocator) noexcept :
+  Allocator allocator,
+  bool own) noexcept :
 AllocationStrategy(name, id),
 mixins::Inspector(),
+m_owns_allocator(own),
```
Comment (davidbeckingsale) [inline src/umpire/strategy/AllocationTracker.cpp]: Can you back this change out?

File: src/umpire/strategy/MonotonicAllocationStrategy.cpp [hunk 2]
```diff
@@ -34,14 +34,23 @@ MonotonicAllocationStrategy::MonotonicAllocationStrategy(
   m_block = m_allocator->allocate(m_capacity);
 }
 
+MonotonicAllocationStrategy::~MonotonicAllocationStrategy()
```
Comment (davidbeckingsale) [inline src/umpire/strategy/MonotonicAllocationStrategy.cpp]: I don't think this function is reachable?

File: src/umpire/strategy/SlotPool.cpp [hunk 3]
```diff
@@ -42,6 +42,24 @@ SlotPool::SlotPool(
   }
 }
 
+SlotPool::~SlotPool()
```
Comment (davidbeckingsale) [inline src/umpire/strategy/SlotPool.cpp]: Don't think this is reachable?

File: src/umpire/tpl/simpool/FixedSizePool.hpp [hunk 4]
```diff
@@ -71,7 +71,8 @@ class FixedSizePool
   ~FixedSizePool() {
     for (struct Pool *curr = pool; curr; ) {
       struct Pool *next = curr->next;
-      MA::deallocate(curr);
+      MA::deallocate(curr->data);
+      IA::deallocate(curr);
```
Comment (davidbeckingsale) [inline src/umpire/tpl/simpool/FixedSizePool.hpp]: 👍

File: src/umpire/strategy/AllocationTracker.cpp (hunk 1 above)
Comment (corbett5) [reply src/umpire/strategy/AllocationTracker.cpp]: Yeah the destructor no longer gets called so it's not necessary. I'll remove it.

File: tests/integration/strategy_tests.cpp [hunk 5]
```diff
@@ -325,7 +338,7 @@ TEST(SizeLimiter, Host)
 {
   auto& rm = umpire::ResourceManager::getInstance();
 
-  auto alloc = rm.makeAllocator<umpire::strategy::SizeLimiter>(
+  auto alloc = rm.makeAllocator<umpire::strategy::SizeLimiter, false>(
```
Comment (davidbeckingsale) [inline tests/integration/strategy_tests.cpp]: Could you back this change out?

File: src/umpire/strategy/MonotonicAllocationStrategy.cpp (hunk 2 above)
Comment (corbett5) [reply src/umpire/strategy/MonotonicAllocationStrategy.cpp]: Correct, none of the `AllocationStrategy` destructors get called anymore. However I thought I'd leave them in there so that no one had to re-implement them. However I'm happy to remove them all.

Comment (davidbeckingsale) [reply src/umpire/strategy/MonotonicAllocationStrategy.cpp]: That would be good. We'll keep your other branch around for future reference.

Comment (corbett5) [reply src/umpire/strategy/MonotonicAllocationStrategy.cpp]: Will do.

File: src/umpire/strategy/FixedPool.inl [hunk 6]
```diff
@@ -77,20 +78,23 @@ FixedPool<T, NP, IA>::FixedPool(
 
 template <typename T, int NP, typename IA>
 FixedPool<T, NP, IA>::~FixedPool() {
-    for (struct Pool *curr = m_pool; curr; ) {
-      struct Pool *next = curr->next;
-      m_allocator->deallocate(curr);
-      m_current_size -= sizeof(T)*m_num_per_pool;
-      curr = next;
-    }
+  for (struct Pool *curr = m_pool; curr; ) {
```
Comment (corbett5) [inline src/umpire/strategy/FixedPool.inl]: @davidbeckingsale I'm going to leave this in because the destructor previously existed.

File: tests/integration/strategy_tests.cpp (hunk 5 above)
Comment (corbett5) [reply tests/integration/strategy_tests.cpp]: This fixes the test on ray, but I'll remove if you want.
https://github.com/LLNL/Umpire/pull/84#discussion_r268299826

Comment (davidbeckingsale) [reply tests/integration/strategy_tests.cpp]: Yup, I'll dig into it.

Comment (corbett5) [reply tests/integration/strategy_tests.cpp]: Should I remove it in the meantime?

Comment (davidbeckingsale) [reply tests/integration/strategy_tests.cpp]: Yes please.

Comment (corbett5) [reply tests/integration/strategy_tests.cpp]: Done.

Comment (corbett5) [reply tests/integration/strategy_tests.cpp]: FYI it passes in a release build, but still fails in debug.

File: src/umpire/ResourceManager.cpp [hunk 7]
```diff
@@ -65,7 +65,6 @@ ResourceManager::getInstance()
 }
 
 ResourceManager::ResourceManager() :
-  m_allocator_names(),
```
Comment (mcfadden8) [inline src/umpire/ResourceManager.cpp]: I'm not sure what ```m_allocator_names``` was used for, but why was it removed?

File: src/umpire/ResourceManager.cpp [hunk 8]
```diff
@@ -316,10 +315,12 @@ void ResourceManager::copy(void* dst_ptr, void* src_ptr, size_t size)
...
   auto src_alloc_record = m_allocations.find(src_ptr);
+  std::ptrdiff_t src_offset = static_cast<char*>(src_ptr) - static_cast<char*>(src_alloc_record->m_ptr);
+  std::size_t src_size = src_alloc_record->m_size - src_offset;
+  
   auto dst_alloc_record = m_allocations.find(dst_ptr);
-
-  std::size_t src_size = src_alloc_record->m_size;
-  std::size_t dst_size = dst_alloc_record->m_size;
+  std::ptrdiff_t dst_offset = static_cast<char*>(dst_ptr) - static_cast<char*>(dst_alloc_record->m_ptr);
+  std::size_t dst_size = dst_alloc_record->m_size - dst_offset;
```
Comment (mcfadden8) [inline src/umpire/ResourceManager.cpp]: This looks good as it appears to fix an offset-related memory operation bug.  Is this considered a leak or a different bug?  Was this found from testing or inspection?

Comment (davidbeckingsale) [reply src/umpire/ResourceManager.cpp]: This is a different bug, but worth fixing.

Comment (mcfadden8) [reply src/umpire/ResourceManager.cpp]: I agree.  But did we should make sure that tests are added to cover this case if they did not exist before.

File: src/umpire/strategy/MonotonicAllocationStrategy.cpp [hunk 9]
```diff
@@ -51,9 +51,7 @@ MonotonicAllocationStrategy::allocate(size_t bytes)
 
 void
 MonotonicAllocationStrategy::deallocate(void* UMPIRE_UNUSED_ARG(ptr))
-{
-  UMPIRE_LOG(Info, "() doesn't do anything");
```
Comment (mcfadden8) [inline src/umpire/strategy/MonotonicAllocationStrategy.cpp]: Why was this removed?  Was it causing a problem?

File: src/umpire/tpl/judy/judyL2Array.h [hunk 10]
```diff
@@ -217,7 +217,7 @@ class judyL2Array {
         bool removeEntry( JudyKey key ) {
             if( 0 != ( _lastSlot = ( vector ** ) judy_slot( _judyarray, ( const unsigned char * ) &key, _depth * JUDY_key_size ) ) ) {
                 // _lastSlot->~vector(); //for use with placement new
-                // delete _lastSlot;
```
Comment (mcfadden8) [inline src/umpire/tpl/judy/judyL2Array.h]: Do we know why the comment above was deleting _lastSlot instead of *_lastSlot?  Also, was this leak in the original source as it was brought over or was it introduced after it was added into Umpire?  The reason I ask is because it would be interesting to understand why it was commented out in the first place.

File: src/umpire/strategy/MonotonicAllocationStrategy.cpp (hunk 9 above)
Comment (davidbeckingsale) [reply src/umpire/strategy/MonotonicAllocationStrategy.cpp]: Good question. @corbett5 ?

**Comment by davidbeckingsale** [comment]
@corbett5 could you address Marty's comments? The additions would be: tests for offsets, and update to changelog regarding the fix.

File: src/umpire/ResourceManager.cpp (hunk 7 above)
Comment (corbett5) [reply src/umpire/ResourceManager.cpp]: I removed it because it wasn't being used for anything.

File: src/umpire/ResourceManager.cpp (hunk 8 above)
Comment (corbett5) [reply src/umpire/ResourceManager.cpp]: @mcfadden8 the `memset` functionality is being exercised but within a different unit test. I'll add two new tests which specifically exercise the offset functionality.

File: src/umpire/strategy/MonotonicAllocationStrategy.cpp (hunk 9 above)
Comment (corbett5) [reply src/umpire/strategy/MonotonicAllocationStrategy.cpp]: This was removed because although `MonotonicAllocationStrategy::deallocate` alone doesn't do anything when wrapped in `AllocationTracker` the `deallocate` method does have side effects. My opinion is that a message like this will discourage the user from calling `deallocate`.

File: src/umpire/tpl/judy/judyL2Array.h (hunk 10 above)
Comment (corbett5) [reply src/umpire/tpl/judy/judyL2Array.h]: I tried to look at the blame but it wasn't very informative. Also as far as I know the judy tpl consists only of `judy.h` and `judy.c`. I wasn't able to find any `judyL2Array` online.

Comment (davidbeckingsale) [reply src/umpire/tpl/judy/judyL2Array.h]: When that happens I blame Marty 😄 

The judy implementation can be found here: https://github.com/mpictor/judy-template

Comment (mcfadden8) [reply src/umpire/tpl/judy/judyL2Array.h]: So this line was uncommented in the original file.  Was in commented out in order to get it to compile?  The last update to the referenced file was 5 years ago.

Comment (corbett5) [reply src/umpire/tpl/judy/judyL2Array.h]: So in the original implementation they have the `delete _lastSlot`
https://github.com/mpictor/judy-template/blob/122e21cb406392d7a3340657baf18cf0a31e370a/src/judyL2Array.h#L210
Which means that it must have been commented out when it was added to Umpire.

Comment (mcfadden8) [reply src/umpire/tpl/judy/judyL2Array.h]: uh-huh..  Okay..  I guess that the research of this is beyond the scope of this particular check-in.  But I think it would be prudent to sanity check the differences on this file from the original implementation as a separate task.

File: src/umpire/strategy/MonotonicAllocationStrategy.cpp (hunk 9 above)
Comment (mcfadden8) [reply src/umpire/strategy/MonotonicAllocationStrategy.cpp]: It doesn't bother me since this was not implemented by me.  But it is a change of behavior and I wanted to make sure @davidbeckingsale saw it.  Since he has already approved this PR, I will drop it.

File: src/umpire/ResourceManager.cpp (hunk 8 above)
Comment (corbett5) [reply src/umpire/ResourceManager.cpp]: Done.

**Comment by corbett5** [comment]
Is there a reason the Travis build seems stalled?

**Comment by rhornung67** [comment]
@corbett5 Todd is looking into this. For some reason we're not getting the number of concurrent builds that we've paid for.  All LLNL projects are experiencing the log jam.
